timelock-impl/src/main/java/com/palantir/atlasdb/timelock/transaction/timestamp/DelegatingClientAwareManagedTimestampService.java

```diff
@@ -84,4 +93,8 @@ public TimestampAndPartition getFreshTimestampForClient(UUID clientIdentifier) {
     public ManagedTimestampService delegate() {
         return delegate;
     }
+
+    private static int calculatePartition(long timestamp) {
+        return (int) timestamp % NUM_PARTITIONS;
```

Comment:
I don't think this is what you meant to write. This is equivalent to

```java
((int) timestamp) % NUM_PARTITIONS
```

which is equivalent to

```java
(timestamp % 2^32) % NUM_PARTITIONS
```

and if NUM_PARTITIONS is a power of 2 then we're fine here, but if e.g. `NUM_PARTITIONS` is 3 then `calculatePartition(2^32) == 0` whereas `2^32 mod 3 = 1`.

In this case, because it's a power of 2 it's fine, but it's not documented it _has_ to be a power of 2. I would write this code as

```
return Ints.checkedCast(timestamp % NUM_PARTITIONS);
```

because I think it follows the order of precedence that you want?

Reply:
```
|  Welcome to JShell -- Version 11.0.1
|  For an introduction type: /help intro

jshell> var x = (1L << 32)
x ==> 4294967296

jshell> var mod = 3
mod ==> 3

jshell> (int) x % mod
$3 ==> 0

jshell> (int) (x % mod)
$4 ==> 1

jshell>
```